# Simple Linear Regression With scikit-learn

These steps are more or less general for most of the regression approaches and implementations. Throughout the rest of the tutorial, you’ll learn how to do these steps for several different scenarios.

## Step 1: Import packages and classes


The first step is to import the package ```numpy``` and the class ```LinearRegression``` from ```sklearn.linear_model```:

In [13]:
import numpy as np
from sklearn.linear_model import LinearRegression

The fundamental data type of NumPy is the array type called numpy.ndarray. The rest of this tutorial uses the term **array** to refer to instances of the type numpy.ndarray.

If you dont know much about Numpy, it is strongly encourage to look it up : https://docs.scipy.org/doc/numpy/user You may google other resources as well.

## Step 2: Provide data

The second step is defining data to work with. The inputs (regressors, 𝑥) and output (response, 𝑦) should be arrays or similar objects. This is the simplest way of providing data for regression:

In [14]:
x = np.array([5, 15, 25, 35, 45, 55]).reshape((-1, 1))
y = np.array([5, 20, 14, 32, 22, 38])

you have two arrays: the input, x, and the output, y. You should call ```.reshape()``` on x because this array must be **two-dimensional**, or more precisely, it must have **one column** and **as many rows as necessary**. That’s exactly what the argument (-1, 1) of .reshape() specifies.


In [15]:
x

array([[ 5],
       [15],
       [25],
       [35],
       [45],
       [55]])

In [16]:
y

array([ 5, 20, 14, 32, 22, 38])

Your array should have the following shape:


x has two dimensions, and x.shape is (6, 1), while y has a single dimension, and y.shape is (6,)

## Step 3: Create a model and fit it

Create a linear regression model and fit it using the existing data. Create an instance of the class LinearRegression, which will represent the regression model:



It's ok if you do not understand the concept of class, can just see it as creating a new object of similar behaviour. In our case, we are looking at Linear Regression behavior.

In [17]:
model = LinearRegression()

I will explain some default values of this class.

It’s time to start using the model. First, you need to call .fit() on model:

In [18]:
model.fit(x, y)

LinearRegression()

With .fit(), you calculate the optimal values of the weights 𝑏₀ and 𝑏₁, using the existing input and output, x and y, as the arguments. In other words, .fit() fits the model. It returns self, which is the variable model itself. That’s why you can replace the last two statements with this one:

In [19]:
model = LinearRegression().fit(x, y)

This statement does the same thing as the previous two. It’s just shorter.

## Step 4: Get results

Get the results to check whether the model works satisfactorily and to interpret it.

You can obtain the coefficient of determination, 𝑅², with .score() called on model:

In [20]:
r_sq = model.score(x, y)
print(f"coefficient of determination: {r_sq}")

coefficient of determination: 0.715875613747954


When you’re applying .score(), the arguments are also the predictor x and response y, and the return value is 𝑅².

The attributes of model are .intercept_, which represents the coefficient 𝑏₀, and .coef_, which represents 𝑏₁:

In [21]:
print(f"intercept: {model.intercept_}")

intercept: 5.633333333333329


In [22]:
print(f"slope: {model.coef_}")

slope: [0.54]


The code above illustrates how to get 𝑏₀ and 𝑏₁. You can notice that .intercept_ is a scalar, while .coef_ is an array.

Note:  In scikit-learn, by convention (https://scikit-learn.org/stable/developers/develop.html#estimated-attributes), a trailing underscore indicates that an attribute is estimated. In this example, .intercept_ and .coef_ are estimated values.

The value of 𝑏₀ is approximately 5.63. This illustrates that your model predicts the response 5.63 when 𝑥 is zero. The value 𝑏₁ = 0.54 means that the predicted response rises by 0.54 when 𝑥 is increased by one.

You’ll notice that you can provide y as a two-dimensional array as well. In this case, you’ll get a similar result. This is how it might look:

In [23]:
new_model = LinearRegression().fit(x, y.reshape((-1, 1)))

In [24]:
print(f"intercept: {new_model.intercept_}")

intercept: [5.63333333]


In [25]:
print(f"slope: {new_model.coef_}")

slope: [[0.54]]


## Predict response

Once you have a satisfactory model, then you can use it for predictions with either existing or new data. To obtain the predicted response, use .predict():

In [26]:
y_pred = model.predict(x)

In [27]:
print(f"predicted response:\n{y_pred}")

predicted response:
[ 8.33333333 13.73333333 19.13333333 24.53333333 29.93333333 35.33333333]


When applying .predict(), you pass the regressor as the argument and get the corresponding predicted response. This is a nearly identical way to predict the response:

In [28]:
y_pred = model.intercept_ + model.coef_ * x

In [29]:
print(f"predicted response:\n{y_pred}")

predicted response:
[[ 8.33333333]
 [13.73333333]
 [19.13333333]
 [24.53333333]
 [29.93333333]
 [35.33333333]]


In this case, you multiply each element of x with model.coef_ and add model.intercept_ to the product.

The output here differs from the previous example only in dimensions. The predicted response is now a two-dimensional array, while in the previous case, it had one dimension.

If you reduce the number of dimensions of x to one, then these two approaches will yield the same result. You can do this by replacing x with x.reshape(-1), x.flatten(), or x.ravel() when multiplying it with model.coef_.

In practice, regression models are often applied for forecasts. This means that you can use fitted models to calculate the outputs based on new inputs:

In [30]:
x_new = np.arange(5).reshape((-1, 1))

In [31]:
x_new

array([[0],
       [1],
       [2],
       [3],
       [4]])

In [32]:
y_new = model.predict(x_new)

In [33]:
y_new

array([5.63333333, 6.17333333, 6.71333333, 7.25333333, 7.79333333])

Here .predict() is applied to the new regressor x_new and yields the response y_new. This example conveniently uses arange() from numpy to generate an array with the elements from 0, inclusive, up to but excluding 5—that is, 0, 1, 2, 3, and 4.

For more information on LinearRegression with scikit-learn. Head to https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html
    

## Multiple Linear Regression With scikit-learn

You can implement multiple linear regression following the same steps as you would for simple regression. The main difference is that your x array will now have two or more columns.

## Steps 1 and 2: Import packages and classes, and provide data

First, you import numpy and sklearn.linear_model.LinearRegression and provide known inputs and output:

In [34]:
import numpy as np
from sklearn.linear_model import LinearRegression

x = [[0, 1], [5, 1], [15, 2], [25, 5], [35, 11], [45, 15], [55, 34], [60, 35]]
y = [4, 5, 20, 14, 32, 22, 38, 43]

x, y = np.array(x), np.array(y)

That’s a simple way to define the input x and output y. You can print x and y to see how they look now:

In [35]:
x

array([[ 0,  1],
       [ 5,  1],
       [15,  2],
       [25,  5],
       [35, 11],
       [45, 15],
       [55, 34],
       [60, 35]])

In [36]:
y

array([ 4,  5, 20, 14, 32, 22, 38, 43])

In multiple linear regression, x is a two-dimensional array with at least two columns, while y is usually a one-dimensional array. This is a simple example of multiple linear regression, and x has exactly two columns.

## Step 3: Create a model and fit it

The next step is to create the regression model as an instance of LinearRegression and fit it with .fit():

In [37]:
model = LinearRegression().fit(x, y)

## Step 4: Get results

In [38]:
r_sq = model.score(x, y)
print(f"coefficient of determination: {r_sq}")
print(f"intercept: {model.intercept_}")
print(f"coefficients: {model.coef_}")


coefficient of determination: 0.8615939258756775
intercept: 5.52257927519819
coefficients: [0.44706965 0.25502548]


You obtain the value of 𝑅² using .score() and the values of the estimators of regression coefficients with .intercept_ and .coef_. Again, .intercept_ holds the bias 𝑏₀, while now .coef_ is an array containing 𝑏₁ and 𝑏₂.

In this example, the intercept is approximately 5.52, and this is the value of the predicted response when 𝑥₁ = 𝑥₂ = 0. An increase of 𝑥₁ by 1 yields a rise of the predicted response by 0.45. Similarly, when 𝑥₂ grows by 1, the response rises by 0.26.

## Step 5: Predict response

In [39]:
y_pred = model.predict(x)

In [40]:
print(f"predicted response:\n{y_pred}")

predicted response:
[ 5.77760476  8.012953   12.73867497 17.9744479  23.97529728 29.4660957
 38.78227633 41.27265006]


The predicted response is obtained with .predict(), which is equivalent to the following:

In [41]:
y_pred = model.intercept_ + np.sum(model.coef_ * x, axis=1)
print(f"predicted response:\n{y_pred}")

predicted response:
[ 5.77760476  8.012953   12.73867497 17.9744479  23.97529728 29.4660957
 38.78227633 41.27265006]


You can predict the output values by multiplying each column of the input with the appropriate weight, summing the results, and adding the intercept to the sum.

You can apply this model to new data as well:

In [42]:
x_new = np.arange(10).reshape((-1, 2))

In [43]:
x_new

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])

In [44]:
y_new = model.predict(x_new)

In [45]:
y_new

array([ 5.77760476,  7.18179502,  8.58598528,  9.99017554, 11.3943658 ])

# Polynomial Regression With scikit-learn

There’s only one extra step: you need to transform the array of inputs to include nonlinear terms such as 𝑥².

## Step 1: Import packages and classes

In [46]:
#the other 2 and
from sklearn.preprocessing import PolynomialFeatures

## Step 2a: Provide data

Keep in mind that you need the input to be a two-dimensional array. That’s why .reshape() is used.

In [47]:
x = np.array([5, 15, 25, 35, 45, 55]).reshape((-1, 1))
y = np.array([15, 11, 2, 8, 25, 32])

## Step 2b: Transform input data



This is the new step that you need to implement for polynomial regression!

As you learned earlier, you need to include 𝑥²—and perhaps other terms—as additional features when implementing polynomial regression. For that reason, you should transform the input array x to contain any additional columns with the values of 𝑥², and eventually more features.

It’s possible to transform the input array in several ways, like using insert() from numpy. But the class PolynomialFeatures is very convenient for this purpose. Go ahead and create an instance of this class:

In [48]:
transformer = PolynomialFeatures(degree=2, include_bias=False)

Before applying transformer, you need to fit it with .fit():

In [49]:
transformer.fit(x)

PolynomialFeatures(include_bias=False)

Once transformer is fitted, then it’s ready to create a new, modified input array. You apply .transform() to do that:

In [50]:
x_ = transformer.transform(x)

That’s the transformation of the input array with .transform(). It takes the input array as the argument and returns the modified array.

Alternate way:
You can also use .fit_transform() to replace the three previous statements with only one:



In [51]:
x_ = PolynomialFeatures(degree=2, include_bias=False).fit_transform(x)

With .fit_transform(), you’re fitting and transforming the input array in one statement. This method also takes the input array and effectively does the same thing as .fit() and .transform() called in that order. It also returns the modified array. This is how the new input array looks:

In [52]:
x_

array([[   5.,   25.],
       [  15.,  225.],
       [  25.,  625.],
       [  35., 1225.],
       [  45., 2025.],
       [  55., 3025.]])

The modified input array contains two columns: one with the original inputs and the other with their squares. You can find more information about PolynomialFeatures here https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html

## Step 3: Create a model and fit it

In [53]:
model = LinearRegression().fit(x_, y)

You should keep in mind that the first argument of .fit() is the modified input array x_ and not the original x.

## Step 4: Get results

In [54]:
r_sq = model.score(x_, y)

In [55]:
print(f"coefficient of determination: {r_sq}")

coefficient of determination: 0.8908516262498564


In [56]:
print(f"intercept: {model.intercept_}")

intercept: 21.372321428571425


In [57]:
print(f"coefficients: {model.coef_}")

coefficients: [-1.32357143  0.02839286]


Again, .score() returns 𝑅². Its first argument is also the modified input x_, not x. The values of the weights are associated to .intercept_ and .coef_. Here, .intercept_ represents 𝑏₀, while .coef_ references the array that contains 𝑏₁ and 𝑏₂.

You can obtain a very similar result with different transformation and regression arguments:

In [58]:
x_new = PolynomialFeatures(degree=2, include_bias=True).fit_transform(x)

In [59]:
x_

array([[   5.,   25.],
       [  15.,  225.],
       [  25.,  625.],
       [  35., 1225.],
       [  45., 2025.],
       [  55., 3025.]])

In [60]:
x_new

array([[1.000e+00, 5.000e+00, 2.500e+01],
       [1.000e+00, 1.500e+01, 2.250e+02],
       [1.000e+00, 2.500e+01, 6.250e+02],
       [1.000e+00, 3.500e+01, 1.225e+03],
       [1.000e+00, 4.500e+01, 2.025e+03],
       [1.000e+00, 5.500e+01, 3.025e+03]])

The first column of x_ contains ones, the second has the values of x, while the third holds the squares of x.

The intercept is already included with the leftmost column of ones, and you don’t need to include it again when creating the instance of LinearRegression. Thus, you can provide fit_intercept=False. This is how the next statement looks:

In [61]:
model = LinearRegression(fit_intercept=False).fit(x_new, y)


The variable model again corresponds to the new input array x_. Therefore, x_ should be passed as the first argument instead of x.

This approach yields the following results, which are similar to the previous case:

In [62]:
r_sq = model.score(x_new, y)

In [63]:
print(f"coefficient of determination: {r_sq}")

coefficient of determination: 0.8908516262498565


In [64]:
print(f"intercept: {model.intercept_}")

intercept: 0.0


In [65]:
print(f"coefficients: {model.coef_}")

coefficients: [21.37232143 -1.32357143  0.02839286]


In [66]:
intercept, coefficients = model.intercept_, model.coef_

You see that now .intercept_ is zero, but .coef_ actually contains 𝑏₀ as its first element. Everything else is the same.

## Step 5: Predict response

If you want to get the predicted response, just use .predict(), but remember that the argument should be the modified input x_ instead of the old x:

In [67]:
y_pred = model.predict(x_new)
print(f"predicted response:\n{y_pred}")

predicted response:
[15.46428571  7.90714286  6.02857143  9.82857143 19.30714286 34.46428571]


You can also notice that polynomial regression yielded a higher coefficient of determination than multiple linear regression for the same problem. At first, you could think that obtaining such a large 𝑅² is an excellent result. It might be.

However, in real-world situations, having a complex model and 𝑅² very close to one might also be a sign of overfitting. To check the performance of a model, you should test it with new data—that is, with observations not used to fit, or train, the model. To learn how to split your dataset into the training and test subsets, can stay tuned next week. 

# Sneakpeek next week:

##  Advanced Linear Regression With statsmodels

## Beyond Linear Regression

The package scikit-learn provides the means for using other regression techniques in a very similar way to what you’ve seen. It contains classes for support vector machines, decision trees, random forest, and more, with the methods .fit(), .predict(), .score(), and so on.